marlowe/src/Language/Marlowe/Client/History.hs

```diff
@@ -0,0 +1,335 @@
+  | Closed
+    {
+      historyInput :: TransactionInput  -- ^ The Marlowe input that was applied.
+    , historyTxId  :: TxId              -- ^ The transaction that resulted from the input being applied.
+    }
+    deriving stock (Eq, Generic, Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+
+-- | Retrieve the history of a role-based Marlowe contract.
+marloweHistory :: AsContractError e
+               => MarloweParams                   -- ^ The Marlowe validator parameters.
+               -> Contract w s e (Maybe History)  -- ^ The original contract and the sequence of redemptions, if any.
+marloweHistory params =
+  do
+    let address = validatorAddress $ smallUntypedValidator params
+    -- The script address contains transactions that have datum.
+    addressTxns <- txsAt address
+    -- When a contract closes, there may be a UTxO at the role address.
+    roleTxns <- txsAt . scriptHashAddress $ rolePayoutValidatorHash params
+    -- When a contract closes, there may be a UTxO at the owner's public key hash address.
+    pkhTxns <- txsAt . flip Address Nothing . PubKeyCredential . unPaymentPubKeyHash =<< ownPaymentPubKeyHash
+    -- TODO: Extract all PKHs from the contract, and query these addresses, too.
+    pure
+      . history params address
+      . nub
+      $ addressTxns <> roleTxns <> pkhTxns
+
+
+-- | Retrieve the history of a role-based Marlowe contract.
+history :: MarloweParams   -- ^ The Marlowe validator parameters.
+        -> Address         -- ^ The Marlowe validator address.
+        -> [ChainIndexTx]  -- ^ The transactions at the Marlowe validator and role validator addresses.
+        -> Maybe History   -- ^ The original contract and the sequence of redemptions, if any.
+history params address citxs =
+  case creationTxOut params address `mapMaybe` citxs of
+    -- If role tokens are minted by the "create" endpoint, then there should only ever be on contract at the address.
+    [creation] -> Just
+                    . Created (tyTxOutRefRef creation) (toMarloweState creation)
+                    $ historyFrom address citxs creation
+    -- Either there is no contract yet, or role tokens have been reused for multiple contracts.
+    _          -> Nothing
+
+
+-- | Construct the sequence of redemptions following from a particular Marlowe transactions.
+historyFrom :: Address             -- ^ The Marlowe validator address.
+            -> [ChainIndexTx]      -- ^ The transactions at the Marlowe validator and role validator addresses.
+            -> MarloweTxOutRef     -- ^ The Marlowe transaction to start from.
+            -> Maybe History       -- ^ The sequence of subsequent redemptions.
+historyFrom address citxs consumed =
+  let
+    consumed' = tyTxOutRefRef consumed
+  in
+    -- The next redemption must consume the input.
+    case (any ((consumed' ==) . txInRef) . S.toList . (^. citxInputs)) `filter` citxs of
+      -- Only one transaction can consume the output of the previous step.
```

Comment:
A bit hard to understand the line of code. Awesome comments, though.

Reply:
I'll expand to make clearer. (Sorry about the pointfree style.)

Reply:
Fixed in cb1fba3cd5ae838b556925ab12230ed5b4b6bc35.